slay/framework.py

```diff
@@ -0,0 +1,513 @@
+            ) from e
+        return
+
+    raise definitions.APIDefinitonError(anno)
+
+
+def _validate_and_describe_endpoint(
+    cls: Type[definitions.ABCProcessor],
+) -> definitions.EndpointAPIDescriptor:
+    if not hasattr(cls, definitions.ENDPOINT_NAME):
+        raise definitions.APIDefinitonError(
+            f"`{cls.__name__}` must have a {definitions.ENDPOINT_NAME}` method."
+        )
+    endpoint_method = getattr(
+        cls, definitions.ENDPOINT_NAME
+    )  # This is the unbound method.
+    if not inspect.isfunction(endpoint_method):
+        raise definitions.APIDefinitonError(
+            f"`{cls.__name__}.{definitions.ENDPOINT_NAME}` must be a method."
+        )
+
+    signature = inspect.signature(endpoint_method)
+    params = list(signature.parameters.values())
+    if params[0].name != definitions.SELF_ARG_NAME:
+        raise definitions.APIDefinitonError(
+            f"`{cls.__name__}.{definitions.ENDPOINT_NAME}` must be a method, i.e. "
+            "with `self` argument."
+        )
+    input_name_and_types = []
+    for param in params[1:]:  # Skip self argument.
+        if param.annotation == inspect.Parameter.empty:
+            raise definitions.APIDefinitonError(
+                f"Inputs of endpoints must have type annotations. Got:\n{signature}"
+            )
+        type_descriptor = definitions.TypeDescriptor(raw=param.annotation)
+        input_name_and_types.append((param.name, type_descriptor))
+
+    if signature.return_annotation == inspect.Parameter.empty:
+        raise definitions.APIDefinitonError(
+            f"Return values of endpoints must be type annotated. Got:\n{signature}"
+        )
+    if get_origin(signature.return_annotation) is tuple:
+        output_types = list(
+            definitions.TypeDescriptor(raw=arg)
+            for arg in get_args(signature.return_annotation)
+        )
+    else:
+        output_types = [definitions.TypeDescriptor(raw=signature.return_annotation)]
+
+    if inspect.isasyncgenfunction(endpoint_method):
+        is_async = True
+        is_generator = True
+    elif inspect.iscoroutinefunction(endpoint_method):
+        is_async = True
+        is_generator = False
+    else:
+        is_async = False
+        is_generator = inspect.isgeneratorfunction(endpoint_method)
+
+    return definitions.EndpointAPIDescriptor(
+        input_names_and_tyes=input_name_and_types,
+        output_types=output_types,
+        is_async=is_async,
+        is_generator=is_generator,
+    )
+
+
+def _get_generic_class_type(var):
+    """Extracts `SomeGeneric` from `SomeGeneric` or `SomeGeneric[T]` uniformly."""
+    origin = get_origin(var)
+    return origin if origin is not None else var
+
+
+def _validate_init_and_get_dependencies(
+    cls: Type[definitions.ABCProcessor],
+) -> Mapping[str, Type[definitions.ABCProcessor]]:
+    signature = inspect.signature(cls.__init__)
+    params = list(signature.parameters.values())
```

Comment:
one thought about this -- what if instead of operating on this raw `params` that we get from the signature, we made a `ValidParams` class that read this, in the __init__ function, and then allowed you to access parameter values in a more type-safe way?